### PR TITLE
feat: add beforeEach and afterEach hooks

### LINKED ISSUE #9: Add `beforeEach` & `afterEach` hooks
Currently, `suite.before` and `suite.after` are tied to the _suite's_ lifecycle. 
This would be added to run before/after each test block within the suite.

```js
test.before(() => {
  console.log('SETUP');
});

test.after(() => {
  console.log('CLEANUP');
});

test.beforeEach(() => {
  console.log('>> BEFORE');
});

test.afterEach(() => {
  console.log('>> AFTER');
});

// ---

test('foo', () => {
  console.log('>>>> TEST: FOO');
});

test('bar', () => {
  console.log('>>>> TEST: BAR');
});

test.run();

// SETUP
// >> BEFORE
// >>>> TEST: FOO
// >> AFTER
// >> BEFORE
// >>>> TEST: BAR
// >> AFTER
// CLEANUP
```

### PR DESCRIPTION
**Pull request by hyanmandian**
Closes #9

### CODE REVIEW

**Comment by lukeed** [comment]
Hey thanks!

This still has some work to go (types and docs), but we also need to thinking about the implementation. I've had this originally too, but it doesn't account for:

* `beforeEach` throwing an Error
* `afterEach` throwing an Error
* `test` throwing an Error & so `afterEach` can't cleanup

I will also ping you from #9 about another discussion point

**Comment by hyanmandian** [comment]
I implemented the idea you suggested in #9 , added missing docs, added missing types, and handle hooks exceptions. Can you help me to find a way to test if it works?

**Comment by lukeed** [comment]
Sorry @hyanmandian 🙈 Can you revert that last commit? It actually causes more issues with timing and control flow than the previous state.

The rest is looking good! I can add the missing bits to what _was_ there.
Thanks!

**Comment by hyanmandian** [comment]
Done @lukeed !

**Comment by lukeed** [comment]
This is looking good, thanks @hyanmandian! I'll be able to add the remaining bits tonight 🙌

File: src/index.js [hunk 1]
```diff
@@ -67,10 +66,13 @@ async function runner(ctx, name) {
 		for (hook of before) await hook();
 		for (test of arr) {
 			try {
+				for (hook of bEach) await hook();
 				await test.handler();
+				for (hook of aEach) await hook();
 				write(PASS);
 				num++;
 			} catch (err) {
+				for (hook of aEach) await hook();
```
Comment (lukeed) [inline src/index.js:75]: Sorry for the delay – I was thinking about it, and this is all the "additional cleanup" that was needed.

Comment (hyanmandian) [reply src/index.js:75]: Nice!

**Pull request merged**